### PR TITLE
Balancing perks and more

### PR DESCRIPTION
**Pull request by taelon875**

### CODE REVIEW

**Comment by zunath** [comment]
> 
> 
> None of the cooking stuff is currently in, forgot this was here it is by no means ready. I think i meant to use the amount of cooking bonus on the item, not the stats.... But yeah the values should be the same for anyone that thinks it.

If this isn't ready yet can you move it out of this PR and put it in a separate branch? Trying to reduce the amount of dead code and this will add to it.

File: SWLOR.Game.Server/Data/Migrations/2020-09-29.2.sql [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    ID,
+    PerkID,
+    Level,
+    Price,
+    Description,
+    SpecializationID
+  )
+VALUES
+  (
+    2319,
```
Comment (zunath) [inline SWLOR.Game.Server/Data/Migrations/2020-09-29.2.sql:44]: I checked on this table and it isn't auto incrementing. Probably lost it when I shifted from SQL Server to MySQL. I'm making a note here because I need to make sure these IDs aren't in use on production. Also going to log a ticket so we can fix auto increments

File: SWLOR.Game.Server/Item/Cooking/Drink.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+using SWLOR.Game.Server.GameObject;
+using SWLOR.Game.Server.Item.Contracts;
+using SWLOR.Game.Server.NWN.Enum;
+using SWLOR.Game.Server.Service;
+
+using SWLOR.Game.Server.ValueObject;
+
+namespace SWLOR.Game.Server.Item.Cooking
+{
+    public class Drink : IActionItem
```
Comment (zunath) [inline SWLOR.Game.Server/Item/Cooking/Drink.cs:13]: As mentioned before, let's move the cooking stuff to a different PR to avoid merging in dead code.

File: SWLOR.Game.Server/Item/ModItem.cs [hunk 3]
```diff
@@ -120,19 +120,19 @@ public float Seconds(NWCreature user, NWItem item, NWObject target, Location tar
 
             if(GetLocalBool(targetItem, "LIGHTSABER") == true)
             {
-                perkBonus = PerkService.GetCreaturePerkLevel(userPlayer, PerkType.SpeedyEngineering) * 0.1f;
+                perkBonus = PerkService.GetCreaturePerkLevel(userPlayer, PerkType.SpeedyCrafting) * 0.1f;
```
Comment (zunath) [inline SWLOR.Game.Server/Item/ModItem.cs:123]: These if statements aren't needed. Just do a single line for the perkBonus.

File: SWLOR.Game.Server/Perk/ForceControl/ForceHeal.cs [hunk 4]
```diff
@@ -37,6 +37,7 @@ public float CooldownTime(NWCreature oPC, float baseCooldownTime, int spellTier)
 
         public void OnImpact(NWCreature creature, NWObject target, int perkLevel, int spellTier)
         {
+            creature.AssignCommand(() => _.ActionPlayAnimation(Animation.LoopingMeditate, 1.0f, 9999));
```
Comment (zunath) [inline SWLOR.Game.Server/Perk/ForceControl/ForceHeal.cs:40]: This animation is going to stop playing as soon as the player moves. Is that ok?

File: SWLOR.Game.Server/Perk/ForceSense/AnimalBond.cs [hunk 5]
```diff
@@ -33,6 +51,136 @@ public float CooldownTime(NWCreature oPC, float baseCooldownTime, int spellTier)
...
         {
+            ApplyEffect(creature, target, spellTier);
+        }
+
+        private void ApplyEffect(NWCreature creature, NWObject target, int spellTier)
+        {
+            Effect effect = _.EffectDominated();
+            int animalbondcreaturerating;
+            float duration = 300f;
+            int challengeratingxp = ((int)_.GetChallengeRating(target));
```
Comment (zunath) [inline SWLOR.Game.Server/Perk/ForceSense/AnimalBond.cs:62]: challengeratingXP -> challengeRatingXP

File: SWLOR.Game.Server/Perk/ForceSense/AnimalBond.cs [hunk 6]
```diff
@@ -33,6 +51,136 @@ public float CooldownTime(NWCreature oPC, float baseCooldownTime, int spellTier)
...
+
+            switch (spellTier)
```
Comment (zunath) [inline SWLOR.Game.Server/Perk/ForceSense/AnimalBond.cs:64]: Please remove this if you aren't going to finish it up. Throw it in another branch and see if someone will pick it up. Code's a mess as it is, no need to make it worse :P

File: SWLOR.Game.Server/Scripts/Delayed/CraftCreateItem.cs [hunk 7]
```diff
@@ -272,9 +278,10 @@ private float CalculateEquipmentBonus(NWPlayer player, SkillType skillType)
                 case SkillType.Engineering: equipmentBonus = effectiveStats.Engineering; break;
                 case SkillType.Fabrication: equipmentBonus = effectiveStats.Fabrication; break;
                 case SkillType.Medicine: equipmentBonus = effectiveStats.Medicine; break;
+                case SkillType.Lightsaber: equipmentBonus = ((sabercraft / 10) + force); break;
```
Comment (zunath) [inline SWLOR.Game.Server/Scripts/Delayed/CraftCreateItem.cs:281]: I'm fine with the atmosphere thing, gives a good bonus. I'm less okay with an equipment bonus being affected by skill. Let's remove this line for now since you're going to be making changes to lightsaber crafting anyway.

File: SWLOR.Game.Server/Service/BaseService.cs [hunk 8]
```diff
@@ -1324,7 +1324,7 @@ public static int CalculateMaxFuel(Guid pcBaseID)
```
Comment (zunath) [inline SWLOR.Game.Server/Service/BaseService.cs:1324]: I think you removed the 0.02f change, but just calling it out real quick to confirm.

File: SWLOR.Game.Server/Service/CraftService.cs [hunk 9]
```diff
@@ -233,13 +233,14 @@ public static float CalculateCraftingDelay(NWPlayer oPC, int skillID)
...
 
             // Identify which perk to use for this skill.
-            if (skillType == SkillType.Weaponsmith) perkType = PerkType.SpeedyWeaponsmith;
-            else if (skillType == SkillType.Armorsmith) perkType = PerkType.SpeedyArmorsmith;
-            else if (skillType == SkillType.Cooking) perkType = PerkType.SpeedyCooking;
-            else if (skillType == SkillType.Engineering) perkType = PerkType.SpeedyEngineering;
-            else if (skillType == SkillType.Fabrication) perkType = PerkType.SpeedyFabrication;
-            else if (skillType == SkillType.Medicine) perkType = PerkType.SpeedyMedicine;
-            else if (skillType == SkillType.Harvesting) perkType = PerkType.SpeedyReassembly;
+            if (skillType == SkillType.Weaponsmith) perkType = PerkType.SpeedyCrafting;
```
Comment (zunath) [inline SWLOR.Game.Server/Service/CraftService.cs:236]: You're killing me, smalls! Drop all these unnecessary if statements broski. Just a single line for perkType (you could even get rid of the variable, but I'd settle for just simplifying the if statements)

File: SWLOR.Game.Server/Service/DeathService.cs [hunk 10]
```diff
@@ -48,10 +48,6 @@ private static void ApplyDurabilityLoss(NWPlayer player)
                 DurabilityService.RunItemDecay(player, equipped, RandomService.RandomFloat(0.10f, 0.50f));
             }
 
-            foreach (var item in player.InventoryItems)
```
Comment (zunath) [inline SWLOR.Game.Server/Service/DeathService.cs:51]: I don't really agree with the change personally but I can live with it.

File: SWLOR.Game.Server/Service/PlayerStatService.cs [hunk 11]
```diff
@@ -476,21 +476,25 @@ public static EffectiveItemStats GetPlayerItemEffectiveStats(NWPlayer player, NW
 
 
                 // Calculate AC
-                if (ItemService.ArmorBaseItemTypes.Contains(item.BaseItemType) ||
-                    ItemService.ShieldBaseItemTypes.Contains(item.BaseItemType))
+                if (ItemService.ArmorBaseItemTypes.Contains(item.BaseItemType))
```
Comment (zunath) [inline SWLOR.Game.Server/Service/PlayerStatService.cs:479]: > According to my testing and math, it should be fine, right now combat is mainly a contest of natural 20's on people that would get affected by this. I still think the numbers may be too high, but i'd rather tweak it over time.

> How would you approach it? Because I know limiting AC gain items in game would be one way but that doesn't affect the legacy stuff. Which is what this mainly hits.

I need to understand the problem better first. Is it just legacy gear or is it stuff people can make today?

File: SWLOR.Game.Server/Service/PlayerStatService.cs [hunk 12]
```diff
@@ -500,6 +504,10 @@ public static EffectiveItemStats GetPlayerItemEffectiveStats(NWPlayer player, NW
 
                     int itemAC = item.CustomAC;
                     itemAC = CalculateAdjustedValue(itemAC, item.RecommendedLevel, skillRankToUse, 0);
+                    if (itemAC > maxAC)
+                        {
```
Comment (zunath) [inline SWLOR.Game.Server/Service/PlayerStatService.cs:508]: Fix the brackets here. They should line up with the if statement

File: SWLOR.Game.Server/Service/WeatherService.cs [hunk 13]
```diff
@@ -218,14 +219,16 @@ private static PlanetaryClimate _GetClimate(NWObject oArea)
                 climate.Heat_Modifier = -5;
                 climate.Humidity_Modifier = -8;
 
+                climate.Snow_Storm = true;
```
Comment (zunath) [inline SWLOR.Game.Server/Service/WeatherService.cs:222]: > Good catch. (Wonder if it even works though....)

I spent some time in here over the weekend. I think there's quite a few things that just don't function - one of them being the thunderstorms. Wouldn't be surprised if there's more to this too, lol.

File: SWLOR.Game.Server/Perk/ForceControl/ForceHeal.cs (hunk 4 above)
Comment (taelon875) [reply SWLOR.Game.Server/Perk/ForceControl/ForceHeal.cs:40]: Yeah its fine, if it becomes an issue I will look for a better option.

File: SWLOR.Game.Server/Service/BaseService.cs (hunk 8 above)
Comment (taelon875) [reply SWLOR.Game.Server/Service/BaseService.cs:1324]: Affirmative.

File: SWLOR.Game.Server/Service/WeatherService.cs (hunk 13 above)
Comment (taelon875) [reply SWLOR.Game.Server/Service/WeatherService.cs:194]: I know when Mith brought this over for us he said he disabled  some of the features. I'm curious if that is what is happening here. It is something I want to look into fixing because having storms and stuff would add a whole new level of depth to module, especially if they can cause detriments to the players that get caught in them. I also want to add DM control to it, but i'm sure this is a lot bigger undertaking then it seems.

File: SWLOR.Game.Server/Perk/ForceAlter/ThrowRock.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+        public int? CooldownCategoryID(NWCreature creature, int? baseCooldownCategoryID, int spellTier)
+        {
+            return baseCooldownCategoryID;
+        }
+
+        public void OnImpact(NWCreature creature, NWObject target, int perkLevel, int spellTier)
+        {
+            int damage;
+            int Mod = creature.WisdomModifier;
```
Comment (zunath) [inline SWLOR.Game.Server/Perk/ForceAlter/ThrowRock.cs:42]: Mod -> mod

File: SWLOR.Game.Server/Perk/ForceAlter/ThrowSaber.cs [hunk 15]
```diff
@@ -84,15 +84,18 @@ public void OnImpact(NWCreature player, NWObject target, int perkLevel, int spel
             int iRange = 15;
             int iCount = 1;
             float fDelay = 0;
+
+            int saberdamage = player.RightHand.DamageBonus;
```
Comment (zunath) [inline SWLOR.Game.Server/Perk/ForceAlter/ThrowSaber.cs:88]: saberdamage -> saberDamage

File: SWLOR.Game.Server/Service/WeatherService.cs (hunk 13 above)
Comment (zunath) [reply SWLOR.Game.Server/Service/WeatherService.cs:194]: I spent a couple hours trying to port it over to FF8 and it's pretty confusing. I don't fully understand how the values affect the weather. Definitely more complicated than I had thought. I probably just need to study it a little longer

File: SWLOR.Game.Server/CustomEffect/DivineCellEffect.cs [hunk 16]
```diff
@@ -21,7 +21,7 @@ public string Apply(NWCreature oCaster, NWObject oTarget, int effectiveLevel)
         public void Tick(NWCreature oCaster, NWObject oTarget, int currentTick, int effectiveLevel, string data)
         {
             if (currentTick % 2 != 0) return;
-            int damage = RandomService.D4(1) + (oCaster.RightHand.DamageBonus / 5);
+            int damage = RandomService.D4(1) + (oCaster.RightHand.DamageBonus / 10);
```
Comment (zunath) [inline SWLOR.Game.Server/CustomEffect/DivineCellEffect.cs:24]: This one is 10 and the others are 8. Intentional?

File: SWLOR.Game.Server/Service/PlayerStatService.cs (hunk 11 above)
Comment (taelon875) [reply SWLOR.Game.Server/Service/PlayerStatService.cs:474]: Both Potentially. But Mainly legacy

File: SWLOR.Game.Server/Data/Migrations/2020-09-29.2.sql (hunk 1 above)
Comment (zunath) [reply SWLOR.Game.Server/Data/Migrations/2020-09-29.2.sql:44]: These are all good on production.

File: SWLOR.Game.Server/Data/Migrations/2020-10-05.1.sql [hunk 17]
```diff
@@ -0,0 +1,9 @@
+﻿Update Perk Set IsActive = 1 Where ID = 175;
+Update Perk Set ExecutionTypeID = 7 Where ID = 175;
+Update PerkLevel Set Description = 'The caster gains 10 Temporary Hit Points Per Tick' Where ID = 2281;
```
Comment (zunath) [inline SWLOR.Game.Server/Data/Migrations/2020-10-05.1.sql:3]: Fix these details since you adjusted the numbers on the perk please.

File: SWLOR.Game.Server/CustomEffect/DivineCellEffect.cs (hunk 16 above)
Comment (taelon875) [reply SWLOR.Game.Server/CustomEffect/DivineCellEffect.cs:24]: Yes, Divine bypasses our DI% with stuff. I

File: SWLOR.Game.Server/Data/Migrations/2020-10-05.1.sql [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+Update PerkLevel Set Description = 'The caster gains 20 Temporary Hit Points Per Tick' Where ID = 2282;
+Update PerkLevel Set Description = 'The caster gains 30 Temporary Hit Points and 5% Concealment Per Tick' Where ID = 2283;
+Update PerkFeat Set ConcentrationTickInterval = 6 Where PerkID = 175;
+Update PerkFeat Set ConcentrationFPCost = 3 Where ID = 97;
```
Comment (zunath) [inline SWLOR.Game.Server/Data/Migrations/2020-10-05.1.sql:7]: Change your where clause to PerkID = 175 AND PerkLevelUnlocked = 1 please. It'll make me feel better being more explicit. Be sure to retest this too. And do it for the next couple lines too.

File: SWLOR.Game.Server/Data/Migrations/2020-10-22.2.sql [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+UPDATE Perk
+Set IsActive = 0
+WHERE ID = 101;
+
+UPDATE Perk
+Set IsActive = 1
+WHERE ID = 175;
+
+Update PerkFeat
```
Comment (zunath) [inline SWLOR.Game.Server/Data/Migrations/2020-10-22.2.sql:25]: The IDs on the following PerkFeats are not right. Use a combination of FeatId AND PerkLevelUnlocked and do a retest.

File: SWLOR.Game.Server/Data/Migrations/2020-10-22.2.sql [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+Set BaseCooldownTime = 30
+Where ID = 16;
+
+Update PerkFeat
+Set ConcentrationTickInterval = 6
+Where PerkID = 175;
+
+Update PerkFeat
+Set ConcentrationFPCost = 2
+Where ID = 97;
```
Comment (zunath) [inline SWLOR.Game.Server/Data/Migrations/2020-10-22.2.sql:55]: Change your where clause to use PerkID and PerkLevelUnlocked please, to be more explicit.

File: SWLOR.Game.Server/Data/Migrations/2020-10-22.3.sql [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+
+
+INSERT INTO CooldownCategory (ID , Name, BaseCooldownTime)
+Values (42, 'Force Crush', 30);
+
+INSERT into Perk (ID, Name, IsActive, BaseCastingTime, Description, PerkCategoryID, CooldownCategoryID, ExecutionTypeID, IsTargetSelfOnly, Enmity, EnmityAdjustmentRuleID, CastAnimationID, ForceBalanceTypeID)
+Values (187, 'Force Crush', 1, 0, 'Damages the Target for 2 plus average of Wisdom and Intelliegence Modifiers', 40, 42, 7, 0, 10, 1, null, 2);
+
+INSERT into PerkLevel (ID, PerkID, Level, Price, Description, SpecializationID)
+Values (2334, 187, 1, 2, 'Damages the Target for 2 Average of Wisdom and Intelligence Modifiers', 0);
```
Comment (zunath) [inline SWLOR.Game.Server/Data/Migrations/2020-10-22.3.sql:61]: These descriptions aren't clear. What is "2 Average" ?

File: SWLOR.Game.Server/Service/PlayerStatService.cs (hunk 11 above)
Comment (zunath) [reply SWLOR.Game.Server/Service/PlayerStatService.cs:474]: I really hate it, but I'll concede to you. We can try it out. Fix the other stuff though :P

**Review by zunath — APPROVED** [review]
